diff -rupNEZbB OriginalSources/usr/include/minix/callnr.h Sources/usr/include/minix/callnr.h
--- OriginalSources/usr/include/minix/callnr.h	2023-05-05 19:36:19.894869333 +0200
+++ Sources/usr/include/minix/callnr.h	2023-05-05 19:36:20.411536018 +0200
@@ -58,8 +58,9 @@
 #define PM_GETEPINFO		(PM_BASE + 45)
 #define PM_GETPROCNR		(PM_BASE + 46)
 #define PM_GETSYSINFO		(PM_BASE + 47)
+#define PM_TRANSFER_MONEY (PM_BASE + 48)
 
-#define NR_PM_CALLS		48	/* highest number from base plus one */
+#define NR_PM_CALLS		49	/* highest number from base plus one */
 
 /*===========================================================================*
  *				Calls to VFS				     *
diff -rupNEZbB OriginalSources/usr/include/minix/config.h Sources/usr/include/minix/config.h
--- OriginalSources/usr/include/minix/config.h	2023-05-05 19:36:19.891536000 +0200
+++ Sources/usr/include/minix/config.h	2023-05-05 19:36:20.414869351 +0200
@@ -95,4 +95,8 @@
 #define NR_PCIBUS 40
 #define NR_PCIDEV 50
 
+/* Money related */
+#define INIT_BALANCE 100
+#define MAX_BALANCE 1000
+
 #endif /* _CONFIG_H */
diff -rupNEZbB OriginalSources/usr/include/unistd.h Sources/usr/include/unistd.h
--- OriginalSources/usr/include/unistd.h	2023-05-05 19:36:19.894869333 +0200
+++ Sources/usr/include/unistd.h	2023-05-06 16:51:12.421304407 +0200
@@ -411,5 +411,7 @@ extern	 int optreset;		/* getopt(3) exte
 extern	 char *suboptarg;	/* getsubopt(3) external variable */
 #endif
 
+int transfermoney(pid_t recipient, int amount);
+
 __END_DECLS
 #endif /* !_UNISTD_H_ */
diff -rupNEZbB OriginalSources/usr/src/include/unistd.h Sources/usr/src/include/unistd.h
--- OriginalSources/usr/src/include/unistd.h	2023-05-05 19:36:20.051536006 +0200
+++ Sources/usr/src/include/unistd.h	2023-05-06 16:51:20.564638025 +0200
@@ -411,5 +411,7 @@ extern	 int optreset;		/* getopt(3) exte
 extern	 char *suboptarg;	/* getsubopt(3) external variable */
 #endif
 
+int transfermoney(pid_t recipient, int amount);
+
 __END_DECLS
 #endif /* !_UNISTD_H_ */
diff -rupNEZbB OriginalSources/usr/src/lib/libc/misc/Makefile.inc Sources/usr/src/lib/libc/misc/Makefile.inc
--- OriginalSources/usr/src/lib/libc/misc/Makefile.inc	2023-05-05 19:36:20.078202673 +0200
+++ Sources/usr/src/lib/libc/misc/Makefile.inc	2023-05-06 16:55:09.474645876 +0200
@@ -7,4 +7,4 @@
 SRCS+=	initfini.c
 
 # for -fstack-protector
-SRCS+=	stack_protector.c
+SRCS+=	stack_protector.c transfermoney.c
diff -rupNEZbB OriginalSources/usr/src/lib/libc/misc/transfermoney.c Sources/usr/src/lib/libc/misc/transfermoney.c
--- OriginalSources/usr/src/lib/libc/misc/transfermoney.c	1970-01-01 01:00:00.000000000 +0100
+++ Sources/usr/src/lib/libc/misc/transfermoney.c	2023-05-08 21:28:25.188930536 +0200
@@ -0,0 +1,15 @@
+#include <stdio.h>
+#include <lib.h>
+
+int transfermoney(pid_t recipient, int amount) {
+  message m;
+  m.m2_l1 = recipient;
+  m.m2_i1 = amount;
+
+  int res = _syscall(PM_PROC_NR, PM_TRANSFER_MONEY, &m);
+
+  if (res < 0)
+    return -1;
+
+  return m.m2_i1;
+}
\ No newline at end of file
diff -rupNEZbB OriginalSources/usr/src/minix/include/minix/callnr.h Sources/usr/src/minix/include/minix/callnr.h
--- OriginalSources/usr/src/minix/include/minix/callnr.h	2023-05-05 19:36:19.951536003 +0200
+++ Sources/usr/src/minix/include/minix/callnr.h	2023-05-05 19:36:20.468202687 +0200
@@ -58,8 +58,9 @@
 #define PM_GETEPINFO		(PM_BASE + 45)
 #define PM_GETPROCNR		(PM_BASE + 46)
 #define PM_GETSYSINFO		(PM_BASE + 47)
+#define PM_TRANSFER_MONEY (PM_BASE + 48)
 
-#define NR_PM_CALLS		48	/* highest number from base plus one */
+#define NR_PM_CALLS		49	/* highest number from base plus one */
 
 /*===========================================================================*
  *				Calls to VFS				     *
diff -rupNEZbB OriginalSources/usr/src/minix/servers/pm/exec.c Sources/usr/src/minix/servers/pm/exec.c
--- OriginalSources/usr/src/minix/servers/pm/exec.c	2023-05-05 19:36:19.971536002 +0200
+++ Sources/usr/src/minix/servers/pm/exec.c	2023-05-05 19:36:20.491536020 +0200
@@ -111,6 +111,10 @@ int do_newexec(void)
 	rmp->mp_frame_addr = (vir_bytes) args.stack_high - args.frame_len;
 	rmp->mp_frame_len = args.frame_len;
 
+  /* Add initial money balance */
+
+  rmp->money_balance = INIT_BALANCE;
+
 	/* Kill process if something goes wrong after this point. */
 	rmp->mp_flags |= PARTIAL_EXEC;
 
diff -rupNEZbB OriginalSources/usr/src/minix/servers/pm/Makefile Sources/usr/src/minix/servers/pm/Makefile
--- OriginalSources/usr/src/minix/servers/pm/Makefile	2023-05-05 19:36:19.971536002 +0200
+++ Sources/usr/src/minix/servers/pm/Makefile	2023-05-05 19:36:20.488202688 +0200
@@ -4,7 +4,7 @@
 PROG=	pm
 SRCS=	main.c forkexit.c exec.c time.c alarm.c \
 	signal.c utility.c table.c trace.c getset.c misc.c \
-	profile.c mcontext.c schedule.c
+	profile.c mcontext.c schedule.c transfermoney.c
 
 DPADD+=	${LIBSYS} ${LIBTIMERS}
 LDADD+=	-lsys -ltimers
diff -rupNEZbB OriginalSources/usr/src/minix/servers/pm/mproc.h Sources/usr/src/minix/servers/pm/mproc.h
--- OriginalSources/usr/src/minix/servers/pm/mproc.h	2023-05-05 19:36:19.971536002 +0200
+++ Sources/usr/src/minix/servers/pm/mproc.h	2023-05-05 19:36:20.488202688 +0200
@@ -70,6 +70,8 @@ EXTERN struct mproc {
   char mp_name[PROC_NAME_LEN];	/* process name */
 
   int mp_magic;			/* sanity check, MP_MAGIC */
+
+  int money_balance;
 } mproc[NR_PROCS];
 
 /* Flag values */
diff -rupNEZbB OriginalSources/usr/src/minix/servers/pm/proto.h Sources/usr/src/minix/servers/pm/proto.h
--- OriginalSources/usr/src/minix/servers/pm/proto.h	2023-05-05 19:36:19.971536002 +0200
+++ Sources/usr/src/minix/servers/pm/proto.h	2023-05-05 20:29:00.994977953 +0200
@@ -89,3 +89,6 @@ struct mproc *find_proc(pid_t lpid);
 int nice_to_priority(int nice, unsigned *new_q);
 int pm_isokendpt(int ep, int *proc);
 void tell_vfs(struct mproc *rmp, message *m_ptr);
+
+/* transfermoney.c */
+int do_transfermoney(void);
diff -rupNEZbB OriginalSources/usr/src/minix/servers/pm/table.c Sources/usr/src/minix/servers/pm/table.c
--- OriginalSources/usr/src/minix/servers/pm/table.c	2023-05-05 19:36:19.971536002 +0200
+++ Sources/usr/src/minix/servers/pm/table.c	2023-05-05 19:36:20.491536020 +0200
@@ -58,5 +58,6 @@ int (* const call_vec[NR_PM_CALLS])(void
 	CALL(PM_EXEC_RESTART)	= do_execrestart,
 	CALL(PM_GETEPINFO)	= do_getepinfo,		/* getepinfo(2) */
 	CALL(PM_GETPROCNR)	= do_getprocnr,		/* getprocnr(2) */
-	CALL(PM_GETSYSINFO)	= do_getsysinfo		/* getsysinfo(2) */
+	CALL(PM_GETSYSINFO)	= do_getsysinfo,	/* getsysinfo(2) */
+  CALL(PM_TRANSFER_MONEY)	= do_transfermoney
 };
diff -rupNEZbB OriginalSources/usr/src/minix/servers/pm/transfermoney.c Sources/usr/src/minix/servers/pm/transfermoney.c
--- OriginalSources/usr/src/minix/servers/pm/transfermoney.c	1970-01-01 01:00:00.000000000 +0100
+++ Sources/usr/src/minix/servers/pm/transfermoney.c	2023-05-08 21:12:13.545563805 +0200
@@ -0,0 +1,28 @@
+#include "pm.h"
+#include "mproc.h"
+
+int do_transfermoney()
+{
+  pid_t recipient = m_in.m2_l1;
+  int amount = m_in.m2_i1;
+
+  register struct mproc *rec_mproc;
+
+  if (!(rec_mproc = find_proc(recipient)))
+    return ESRCH;
+
+  if (rec_mproc->mp_parent == who_p || &mproc[mp->mp_parent] == rec_mproc)
+    return EPERM;
+
+  if (amount < 0 ||
+      mp->money_balance < amount ||
+      rec_mproc->money_balance > MAX_BALANCE - amount)
+    return EINVAL;
+
+  mp->money_balance -= amount;
+  rec_mproc->money_balance += amount;
+
+  mp->mp_reply.m2_i1 = mp->money_balance;
+
+  return OK;
+}
